net/src/vxlan/vni.rs

```diff
@@ -0,0 +1,206 @@
+#[repr(transparent)]
+pub struct Vni(NonZero<u32>);
+
+impl Vni {
+    /// The minimum legal [`Vni`] value (1).
+    pub const MIN: u32 = 1;
+    /// The maximum legal [`Vni`] value (2<sup>24</sup> - 1).
+    pub const MAX: u32 = 0x00_FF_FF_FF;
+    /// First value which is too large to be a legal [`Vni`]
+    #[allow(unused)] // used in test suite
+    const TOO_LARGE: u32 = Vni::MAX + 1;
+
+    /// Create a new [`Vni`] from a `u32`.
+    ///
+    /// # Errors
+    ///
+    /// Returns an [`InvalidVni`] error if the value is 0 or greater than [`Vni::MAX`].
+    pub fn new_checked(vni: u32) -> Result<Vni, InvalidVni> {
+        match NonZero::<u32>::new(vni) {
+            None => Err(InvalidVni::ReservedZero),
+            _ if vni > Vni::MAX => Err(InvalidVni::TooLarge(vni)),
+            Some(vni) => Ok(Vni(vni)),
+        }
+    }
+
+    /// Get the value of the [`Vni`] as a `u32`.
+    #[must_use]
+    pub fn as_u32(self) -> u32 {
+        self.0.get()
+    }
+}
+
+impl AsRef<NonZero<u32>> for Vni {
+    #[must_use]
+    fn as_ref(&self) -> &NonZero<u32> {
+        &self.0
+    }
+}
+
+/// Errors that can occur when converting a `u32` to a [`Vni`]
+#[must_use]
+#[derive(Copy, Clone, Debug, PartialEq, Eq, Hash, PartialOrd, Ord, thiserror::Error)]
+#[cfg_attr(feature = "serde", derive(serde::Deserialize, serde::Serialize))]
+pub enum InvalidVni {
+    /// Zero is not a legal Vni in many EVPN / VXLAN implementations.  Don't use it.
+    #[error("Zero is not a legal Vni")]
+    ReservedZero,
+    /// This error type contains the (illegal) value used to attempt creation of a [`Vni`].
+    /// The max legal value is found in [`Vni::MAX`].
+    #[error("The value {0} is too large to be a Vni (max is {MAX})", MAX = Vni::MAX)]
+    TooLarge(u32),
+}
+
+impl From<Vni> for u32 {
+    fn from(vni: Vni) -> u32 {
+        vni.as_u32()
+    }
+}
+
+impl TryFrom<u32> for Vni {
+    type Error = InvalidVni;
+
+    fn try_from(vni: u32) -> Result<Vni, Self::Error> {
+        Vni::new_checked(vni)
+    }
+}
+
+#[cfg(any(test, feature = "arbitrary"))]
+mod contract {
+    use crate::vxlan::Vni;
+    use arbitrary::{Arbitrary, Unstructured};
+
+    impl<'a> Arbitrary<'a> for Vni {
+        fn arbitrary(u: &mut Unstructured<'a>) -> arbitrary::Result<Self> {
+            let raw: u32 = u.arbitrary::<u32>()? & Vni::MAX;
+            if raw == 0 {
+                Ok(Vni::new_checked(1).unwrap_or_else(|e| unreachable!("{e:?}")))
+            } else {
+                Ok(Vni::new_checked(raw).unwrap_or_else(|e| unreachable!("{e:?}")))
+            }
+        }
+
+        fn size_hint(_depth: usize) -> (usize, Option<usize>) {
+            (size_of::<Vni>(), Some(size_of::<Vni>()))
+        }
+    }
+}
+
+#[cfg(test)]
+#[allow(clippy::unwrap_used, clippy::expect_used)] // valid in test code
+mod test {
+    use super::*;
+
+    #[test]
+    fn zero_is_not_a_legal_vni() {
+        assert_eq!(Vni::new_checked(0).unwrap_err(), InvalidVni::ReservedZero);
```

Comment:
We should not generate (de novo) a VNI of 0, but we should accept it if the RFC says it is a legal VNI.  Do not change this, create an issue for it.

Reply:
Having thought about this more, I'm not sure it makes sense to accept a VNI you can't respond to.  The VXLAN tunnel would be fundamentally broken if only one side of it worked.  That said, I agree with you in principal.  I'm happy to make a ticket where we can discuss tho